src/alluvium/avro.py

```diff
@@ -0,0 +1,251 @@
+"""
+Defines the schema you expect from your CDC events.
+
+Provides synctactic sugar for Any types and Either Types.
+"""
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, List, Optional, Union
+
+
+class InvalidSchemaException(Exception):
+    def __init__(self, field_type, message, name=None):
+        self.error_field_type = field_type
+        self.name = name
+        super(InvalidSchemaException, self).__init__(
+            self, "Error with field <{} {}>. {}".format(field_type, name, message)
+        )
+
+
+class AvroPrimitiveType(Enum):
+    NULL = "null"
+    BOOLEAN = "boolean"
+    INTEGER = "int"
+    LONG = "long"
+    FLOAT = "float"
+    DOUBLE = "double"
+    BYTES = "bytes"
+    STRING = "string"
+
+
+class AvroComplexType(Enum):
+    RECORD = "record"
+    ENUM = "enum"
+    ARRAY = "array"
+    MAP = "map"
+    UNIONS = "unions"
+    FIXED = "fixed"
+
+
+class AvroCustomType(Enum):
+    ANY = "any"
+
+
+def resolve_primitive_type(avro_type: Union[AvroPrimitiveType, AvroCustomType]):
+    if isinstance(avro_type, AvroPrimitiveType):
+        return avro_type.value
+    if avro_type == AvroCustomType.ANY:
+        return AvroUnionSchema([primitive for primitive in AvroPrimitiveType]).avro_schema()
+    raise ValueError("Invalid base type provided")
+
+
+class FieldOrdering(Enum):
+    ASCENDING = "ascending"
+    DESCENDING = "descending"
+    IGNORE = "ignore"
+
+
+class AvroSchema(ABC):
+    def __init__(self, avro_type: Union[AvroPrimitiveType, AvroComplexType]):
+        self.type = avro_type
+
+    @abstractmethod
+    def avro_schema(self):
+        pass
+
+
+class AvroRecordField(object):
+    def __init__(
+        self,
+        name: str,
+        avro_type: Union[AvroSchema, AvroPrimitiveType],
```

Comment:
Is there any reason AvroComplexType is not included in the avro_type for a AvroRecordField? The AvroSchema itself is Union[AvroPrimitiveType, AvroComplexType]. There might be a redundancy here.

Reply:
Ah so `AvroRecordField` is a special exception here. The `AvroRecordSchema` is a Avro Schema. However, the field is a special object that is not necessarily a schema. The reason for that is to support nesting. Here is an example:

```
nested_schema = AvroRecordSchema(
    name='nested', 
    fields=[
        AvroRecordField(name='nested_field', avro_type=AvroArrayType(name='foo', items= AvroPrimitiveType.INT)),
    ]
```

Normally in schema's, the avro_type is a string. However, in fields, the avro type is actually another schema which is why you have that typing divergence. You are totally right that this could be treated as a schema but I just didn't know how to resolve nesting with labels without confusing the semantics.